pkg/controller/volume/persistentvolume/scheduler_binder_test.go

```diff
@@ -1383,6 +1395,47 @@ func TestBindPodVolumes(t *testing.T) {
 			},
 			shouldFail: true,
 		},
+		"node-affinity-check-fails": {
+			binding:   binding1aBoundSelectedNode,
+			cachedPV:  pvNode1a,
+			cachedPVC: selectedNodePVC,
+			delayFunc: func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim) {
+				// Update PVC to be fully bound to a PV with a different node
+				newPVC := pvc.DeepCopy()
+				newPVC.ResourceVersion = "100"
+				newPVC.Spec.VolumeName = pvNode2.Name
+				metav1.SetMetaDataAnnotation(&newPVC.ObjectMeta, annBindCompleted, "yes")
+
+				// Update the pod binding cache with the new PVC
+				bindingsCache := testEnv.binder.GetBindingsCache()
```

Comment:
I think you don't need to update the bindings cache.  We only use the pvc.name and namespace from the pod bindings cache, but then in `checkBindings`, we will actually look up the pvc object from the pvc assume cache.  Trying to mirror the "binding succeeds after time" test case as close as possible.

Reply:
> in checkBindings, we will actually look up the pvc object from the pvc assume cache

Right, `checkBindings` looks up the object from the assume cache, however, `checkBoundClaims` doesn't do that currently; it relies on the list of `v1.PersistentVolumeClaim` passed in to it [1].

In fact, we may need to update the binding cache in other tests as well. For example, `checkBindings` is only succeeding in _binding succeeds after time_ because both `claimsToProvision` and `bindings` are empty lists [2] (since they are not present in the pod binding cache).

Maybe am I missing something?

[1] https://github.com/bertinatto/kubernetes/blob/2e1ef482336791e5a984729227153bf756a47410/pkg/controller/volume/persistentvolume/scheduler_binder.go#L320
[2] https://github.com/bertinatto/kubernetes/blob/2e1ef482336791e5a984729227153bf756a47410/pkg/controller/volume/persistentvolume/scheduler_binder.go#L397

Reply:
I want to use the assume cache as the source of truth for API objects, not the pod binding cache, which is only meant to store binding/provisioning decisions wrt the the pvc and pv names.

Instead of calling `checkBoundClaims` directly, can you get the PVC object from the assume cache like [this](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/controller/volume/persistentvolume/scheduler_binder.go#L409), and then call `CheckNodeAffinity` directly?

Reply:
`claimsToProvision` [is always empty](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/controller/volume/persistentvolume/scheduler_binder.go#L294) during unit tests execution (because we don't update the pod binding cache), so I can't get any PVC objects from the assume cache because I don't know their names.

I could get the PVC/PV names from [bindings instead](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/controller/volume/persistentvolume/scheduler_binder.go#L293), but that sounds incorrect. Here's [sketch](https://gist.github.com/bertinatto/7b4d870c6c1dd14adf6301c4155db77a) of that.

Reply:
Can you modify the test case to also populate the Pod binding cache with initial PVC objects?